### PR TITLE
GYB format version 3

### PR DESCRIPTION
**Pull request by jpcima**
#86

- support selecting the file version to save
- support reading GYB v3

TODO:
- writing GYB v3

### CODE REVIEW

File: src/FileFormats/format_m2v_gyb.cpp [hunk 1]
```diff
@@ -337,22 +455,157 @@ FfmtErrCode M2V_GYB::saveFileWithVersion(QString filePath, FmBank &bank, uint8_t
     return FfmtErrCode::ERR_OK;
 }
 
-int M2V_GYB::formatCaps() const
+FfmtErrCode Basic_M2V_GYB::saveFileVersion3(QFile &file, FmBank &bank)
 {
-    return (int)FormatCaps::FORMAT_CAPS_EVERYTHING;
+    //TODO: implement me version 3
+    
+    return FfmtErrCode::ERR_BADFORMAT;
```
Comment (Wohlstand) [inline src/FileFormats/format_m2v_gyb.cpp]: Here is `ERR_NOT_IMPLEMENTED` error you would to put here :thinking:

File: src/FileFormats/format_m2v_gyb.h [hunk 2]
```diff
@@ -21,22 +21,74 @@
...
     int         formatCaps() const override;
     QString     formatName() const override;
-    QString     formatExtensionMask() const override;
     BankFormats formatId() const override;
+};
 
-protected:
-    FfmtErrCode saveFileWithVersion(QString filePath, FmBank &bank, uint8_t version);
+/**
+ * @brief Writer of GYB File Format, version 1
```
Comment (Wohlstand) [inline src/FileFormats/format_m2v_gyb.h]: did you meant Version 2? :smirk:

File: src/FileFormats/format_m2v_gyb.h [hunk 3]
```diff
@@ -21,22 +21,74 @@
...
+public:
+    FfmtErrCode saveFile(QString filePath, FmBank &bank) override;
+    int         formatCaps() const override;
+    QString     formatName() const override;
+    BankFormats formatId() const override;
+};
+
+
+/**
+ * @brief Writer of GYB File Format, version 1
```
Comment (Wohlstand) [inline src/FileFormats/format_m2v_gyb.h]: .... and 3?

**Comment by jpcima** [comment]
You're right. I try to get these copy-pastes correct :smile:
I'll clean up soon and I expect to finish GYBv3 support sometime in this afternoon.

File: src/FileFormats/format_m2v_gyb.cpp (hunk 1 above)
Comment (jpcima) [reply src/FileFormats/format_m2v_gyb.cpp]: No that would be actually `ERR_NOT_IMLEMENTED`.
This enum has a typo in the name.

Comment (Wohlstand) [reply src/FileFormats/format_m2v_gyb.cpp]: So, it's need to fix it... My fail...

**Comment by Wohlstand** [comment]
Here is a patch to fix this macro error, please check is it compatible with your state? I hope it should, then, ship it together with this PR:
[0001-Fixed-a-typo-in-a-macro.patch.txt](https://github.com/Wohlstand/OPN2BankEditor/files/3910821/0001-Fixed-a-typo-in-a-macro.patch.txt)

File: src/FileFormats/format_m2v_gyb.cpp [hunk 4]
```diff
@@ -337,22 +455,157 @@ FfmtErrCode M2V_GYB::saveFileWithVersion(QString filePath, FmBank &bank, uint8_t
...
+    uint8_t version = header[2];
+    switch(version)
+    {
+    case 1:
+    case 2:
+        return loadFileVersion1Or2(file, bank, version);
+    case 3:
+        return loadFileVersion3(file, bank);
+    default:
+        return FfmtErrCode::ERR_BADFORMAT;
```
Comment (Wohlstand) [inline src/FileFormats/format_m2v_gyb.cpp]: You can use `ERR_UNSUPPORTED_FORMAT` to specify that it's version is unsupported. Means, format is correct, but has a version that is not supported :eyes:

**Comment by jpcima** [comment]
I am able to export some banks to mid2smps and play with them, it seems fine.

**Pull request merged**